Add error handling to MPAS_stream_mgr_remove_alarm and remove redundant logging

amstokely opened the pull request:
This PR introduces error checking to the `MPAS_stream_mgr_remove_alarm` subroutine in `mpas_stream_manager.F`, ensuring that errors returned by the `mpas_stream_list_remove` subroutine are correctly handled. Previously, specific logic paths (e.g., passing an incorrect direction to `MPAS_stream_mgr_remove_alarm`) would log an error without affecting the state of the code. This issue has been addressed by adding the necessary error handling, which now ensures the correct error codes are set and logged when errors occur.

Additionally, redundant error logging in the `mpas_stream_list_module` (specifically in the `mpas_stream_list_remove` subroutine) has been removed, as the newly added error checking now handles and logs these errors correctly.

### Testing:
The changes can be verified by running the test core. The stream tests, which tested this error checking, in commit [7bc8ea0f518d59a9f30557446b70bb656d452e5f](https://github.com/MPAS-Dev/MPAS-Model/tree/7bc8ea0f518d59a9f30557446b70bb656d452e5f) (the first commit in this PR), are now passing, confirming that the issue has been resolved.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,197 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+   !> \details This subroutine calls the `test_remove_alarm` subroutine,
+   !>          passing the `MPAS_STREAM_INPUT` as the stream type to test
+   !>          the functionality of removing an input alarm from a stream.
+   !
```
It looks like the commit in the lines above should go with the `test_remove_input_alarm` subroutine, rather than with the `test_core_streams_test` subroutine in the line immediately below.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -35,7 +35,7 @@ module test_core_streams
    !
    !-----------------------------------------------------------------------
-   subroutine test_core_streams_test(domain, threadErrs, ierr)
+   subroutine test_read_write_real_streams(domain, threadErrs, ierr)
```
It would be good to update the routine name on line 22 in the comment block above.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,197 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+      end if
+
+      call mpas_log_write('Testing removing an input alarm from a stream.')
+      call test_remove_input_alarm(domain, threadErrs, ierr)
```
Did you intend for the third argument to be `ierr3` in this call?

amstokely replied:
I did not...good catch!

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,197 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+         call mpas_log_write('Removing input alarm test: FAILURE', MPAS_LOG_ERR)
+      end if
+
+      ierr = abs(ierr1) + abs(ierr2) + abs(ierr3)
```
Rather than having a new `ierrX` variable for every test, I think we could just use a local `ierr` variable and immediately add its absolute value to the return `ierr` after each test call.

mgduda commented:
@amstokely Rather than combining the changes to `mpas_stream_manager.F` and `mpas_stream_list.F` in the same commit, it may be better to split those changes into two commits: one to ensure complete error handling in the stream manager for all calls to the `mpas_stream_list` module that might return an error code, and a second one to then remove the redundant error logging in the `mpas_stream_list` module.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,205 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+   !>                    and clock.
+   !> \param ierr        The error code that indicates the result of the test.
+   !> \param streamType  The type of the stream being tested (input or output).
+   !
```
As on line 37, can we add a line of hyphens to terminate these documentation blocks? It would be nice to maintain consistent style.

mgduda replied:
It looks like this documentation block still needs a line of hyphens.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,205 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+      integer, dimension(:), intent(out) :: threadErrs
+      integer, intent(out) :: ierr
+
+      integer, dimension(3) :: test_status
```
What's the rationale for making `test_status` an array? Would it be better to just follow what you've done in `test_remove_alarm`, and to have a local scalar to capture the return error code from each test routine, and to add that to the overall `ierr` return code after each test?

amstokely replied:
I initially thought this approach would allow the error code to be set to zero in one place. However, since the `ierr` argument is `intent(out)`, it must be initialized in the test anyway, so using an array doesn't offer any advantage. I'll update this in my next commit.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,205 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+   !> \param ierr        The error code that indicates the result of the test.
+   !> \param streamType  The type of the stream being tested (input or output).
+   !
+   subroutine test_remove_alarm(domain, ierr, streamType)
```
Just as a style suggestion, we almost invariably let the return error code (`ierr`) be the last argument to a subroutine.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,205 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+      call mpas_log_write('Testing reading/writing real-valued streams')
+
+      call test_read_write_real_streams(domain, threadErrs, test_status(1))
+
```
Here and on line 567, removing the blank lines would give more consistent whitespace with the blocks of code for alarm removal tests below.

mgduda commented on src/core_test/mpas_test_core_streams.F:
```diff
@@ -391,6 +391,208 @@ subroutine test_core_streams_test(domain, threadErrs, ierr)
+      integer, intent(out) :: ierr
+
+      call test_remove_alarm(domain, MPAS_STREAM_OUTPUT, ierr)
+   end subroutine test_remove_output_alarm
```
Since this routine is just calling `test_remove_alarm` with a specific `streamType` argument and nothing more, what do you think about just eliminating this routine as well as `test_remove_input_alarm` and calling `test_remove_alarm` directly from `test_core_streams_test`?

amstokely replied:
Works for me!

Commit:
Added unit test for error handling in MPAS_stream_mgr_remove_alarm.

Tests fail when removing an input alarm from an output list or an output
alarm from an input list due to the absence of proper error handling in
MPAS_stream_mgr_remove_alarm in mpas_stream_manager.F. The test checks
for an error when attempting to remove an alarm from the opposite list,
but the error handling is missing, causing the test to fail.

Commit:
Handle errors from MPAS_stream_list_remove in MPAS_stream_mgr_remove_alarm

Previously, some logic paths (e.g., passing the wrong direction to
MPAS_stream_mgr_remove_alarm) would log an error without affecting the
code's execution state. The fix can be verified by running the test
core, where the previously failing stream tests now pass.

Commit:
Remove error logging macros from mpas_stream_list.F

Error logging is no longer necessary in mpas_stream_list.F, as error codes
and logging are now handled in mpas_stream_manager.F. The related macros
have been commented out.

mgduda commented:
In the commit message for 44d15ed2, you state:
```
Tests fail when adding input alarm to an output list or output alarm to an
input list due to absence of error handling.
```
But it looks to me like the failure happens when we try to remove an alarm that doesn't exist. Is that right? If so, can you update the commit message?

If you'll be amending commit 44d15ed2, it might be nice to declare the dummy arguments to the `test_remove_alarm` routine in the order that they appear in the argument list.

mgduda commented:
Also, if you'll be reworking the commit history to address the above comment, can you also give another try at keeping the one-line summary in each commit to one line (and ~80 characters or less)? For example, I think the summary lines for 0d4a7fe6
```
Added error checking to MPAS_stream_mgr_remove_alarm to handle errors
returned by mpas_stream_list_remove.
```
could be shortened to
```
Handle errors returned by MPAS_stream_list_remove in MPAS_stream_mgr_remove_alarm
```
.
